### ISSUE TITLE
Reference to a custom type array becomes null in @sanity-typed/client fetch result

### ISSUE THREAD

**Issue by jschuur** [saiichihashimoto/sanity-typed#251]
I have a field that is an array of references to a custom type:

```ts
    defineField({
      name: 'projects',
      title: 'Projects',
      type: 'array',
      group: 'content',
      of: [
        defineArrayMember({
          type: 'reference',
          to: [{type: 'project'} as const],
        }),
      ],
    }),
```
`project` is set up like this:

```ts
import {defineArrayMember, defineField, defineType} from '@sanity-typed/types'

export default defineType({
  name: 'project',
  title: 'Projects',
  type: 'document',
  fields: [
    defineField({
      name: 'name',
      title: 'Name',
      type: 'string',
      validation: (Rule) => Rule.required(),
    }),
    defineField({
      name: 'listProject',
      title: 'List Project?',
      description: 'Temporarily hide a project.',
      type: 'boolean',
      initialValue: true,
      validation: (Rule) => Rule.required(),
    }),
    defineField({
      name: 'url',
      title: 'URL',
      type: 'string',
      validation: (Rule) => Rule.required(),
    }),
    defineField({
      name: 'repoUrl',
      title: 'Project Git Repository URL',
      type: 'string',
    }),
    defineField({
      name: 'description',
      title: 'Description',
      type: 'blockContent',
      description: 'Content with markup shown on the Projects page.',
    }),
    defineField({
      name: 'summary',
      title: 'Summary',
      type: 'text',
      rows: 3,
      description: 'Shorter description without markup for featured content cards.',
    }),
    defineField({
      name: 'releasedAt',
      title: 'Released At',
      type: 'date',
    }),
    defineField({
      name: 'techStack',
      title: 'Tech Stack',
      type: 'array',
      of: [
        defineArrayMember({
          type: 'reference',
          to: [{type: 'tech'} as const],
        }),
      ],
    }),
  ],
  preview: {
    select: {
      name: 'name',
      url: 'url',
      listProject: 'listProject',
      releasedAt: 'releasedAt',
    },
    prepare: ({name, url, listProject, releasedAt}) => ({
      title: name,
      subtitle:
        (listProject ? '' : 'HIDDEN: ') +
        url +
        (releasedAt ? ' (' + releasedAt.split('Z')[0] + ')' : ''),
    }),
  },
})
```

When I go to query it, however, I get a type of `projects: null`

```ts
import type { SanityValues } from "@joostschuur/cms";
import { createClient } from "@sanity-typed/client";

const projectId = process.env.SANITY_PROJECT_ID;
const dataset = process.env.SANITY_DATASET || "production";

const client = createClient<SanityValues>()({
  projectId,
  dataset,
  useCdn: true,
  apiVersion: "2023-05-23",
});

export const getProjectsPage = async () =>
  (await client.fetch(`*[_type == "projectspage"]{...,projects[]->}`))?.[0];
```

<img width="848" alt="CleanShot 2023-08-27 at 22 37 35@2x" src="https://github.com/saiichihashimoto/sanity-typed/assets/92457/d82c4f5d-83e4-49ea-b89c-d2d88680e81e">

**Comment by jschuur** [comment]
Full WIP code is at https://github.com/jschuur/astro-sanity (lots broken, but it might help with this issue).

**Comment by saiichihashimoto** [comment]
I believe this has to do with getting the [proper traversal cases tested](https://github.com/saiichihashimoto/sanity-typed/issues/198). You're doing a projection on an array type and you should receive an array of the projected types, but you're probably not.

Right now, I'm working on getting all of the `groq` type test cases also mirrored against equivalent `groq-js` test cases, to see if my assumptions about these tests were correct. I can already see they aren't, notably projections on an array type are broken. I'll be pushing out a PR with those tests soon, and then hopefully I can build a fix on it.

**Comment by saiichihashimoto** [comment]
Can you try this with `@sanity-typed/client` at `1.1.3`? Let me know if it works!

**Comment by saiichihashimoto** [comment]
(or beyond that type, I've been trying to improve things)

**Comment by saiichihashimoto** [comment]
@LBBO discovered in a different ticket that projection of optional values was always giving a `null`, so I have a fix for that in #265. Not sure if that will solve this specific issue so, once it's merged and released, let me know if that version solves this issue.

**Comment by LBBO** [comment]
Yep, works like a charm now. Thanks for the quick fix!
